qa/workunits/windows/py_tests/internal/task_group.py

```diff
@@ -0,0 +1,62 @@
+from concurrent import futures
+import logging
+import threading
+
+
+LOG = logging.getLogger()
+
+
+class TaskGroup(object):
+    def __init__(self, max_workers=1, stop_on_error=True):
+        self._executor = futures.ThreadPoolExecutor(max_workers=max_workers)
+        self._lock = threading.Lock()
+
+        self.errors = 0
+        self.completed = 0
+        self.pending = 0
+
+        self.stopped = False
+        self.stop_on_error = stop_on_error
+
+        self._submitted_tasks = []
+
+    def _wrap_task(self, task):
+        def wrapper():
+            if self.stopped:
+                return
+
+            try:
+                task()
+            except KeyboardInterrupt:
+                LOG.warning("Received Ctrl-C.")
+                self.stopped = True
+            except Exception as ex:
+                if self.stop_on_error:
+                    self.stopped = True
+                with self._lock:
+                    self.errors += 1
+                    LOG.exception(
+                        "Task exception: %s. Total exceptions: %d",
+                        ex, self.errors)
+            finally:
+                with self._lock:
+                    self.completed += 1
+                    self.pending -= 1
+                    LOG.info("Completed tasks: %d. Pending: %d",
+                             self.completed, self.pending)
+
+        return wrapper
+
+    def submit(self, task):
+        task_wrapper = self._wrap_task(task)
+
+        submitted_task = self._executor.submit(task_wrapper)
+        with self._lock:
+            self.pending += 1
+            self._submitted_tasks.append(submitted_task)
```

Comment:
Indeed, since ``join`` and ``submit`` are expected to be called from the same thread, we don't need to protect ``self._submitted_tasks``.

As for ``self.stopped``, getting/setting a Python boolean is thread safe as far as I know. Of course, we can be extra cautious and protect it using ``self._lock``.